crates/typst/src/image.rs

```diff
@@ -306,55 +306,100 @@ fn decode_svg(
-        if let Some(success) = referenced.get(&family) {
-            return success.clone();
+        if let Some(success) = font_cache.get(&family) {
+            return Some(success.clone());
+        }
+
+        let fonts = loader.load(&family);
+        let usvg_family = fonts.iter().find_map(|font| {
+            font.find_name(ttf_parser::name_id::TYPOGRAPHIC_FAMILY)
+                .or_else(|| font.find_name(ttf_parser::name_id::FAMILY))
+                .map(Into::<EcoString>::into)
+        });
+
+        let font_data = FontData { usvg_family: usvg_family?, fonts };
+        font_cache.insert(family, font_data.clone());
+        Some(font_data)
+    };
+
+    // Loads a font family into the fontdb database.
+    let mut load_into_db = |font_data: &FontData| {
+        if loaded.contains_key(&font_data.usvg_family) {
+            return;
         }
 
         // We load all variants for the family, since we don't know which will
         // be used.
-        let mut name = None;
-        for font in loader.load(&family) {
+        for font in &font_data.fonts {
             let source = Arc::new(font.data().clone());
             fontdb.load_font_source(fontdb::Source::Binary(source));
-            if name.is_none() {
-                name = font
-                    .find_name(ttf_parser::name_id::TYPOGRAPHIC_FAMILY)
-                    .or_else(|| font.find_name(ttf_parser::name_id::FAMILY))
-                    .map(Into::into);
-            }
         }
 
-        referenced.insert(family, name.clone());
-        name
+        loaded.insert(font_data.usvg_family.clone(), ());
     };
 
-    // Load fallback family.
-    let mut fallback_usvg_compatible = None;
-    if let Some(family) = loader.fallback_family() {
-        fallback_usvg_compatible = load(family);
-    }
+    let fallback_fonts = loader
+        .fallback_families()
+        .iter()
+        .filter_map(|family| load(family.as_str()))
+        .collect::<EcoVec<_>>();
 
-    // Find out which font families are referenced by the SVG.
+    // Determine the best font for each text node.
     traverse_svg(&tree.root, &mut |node| {
         let usvg::NodeKind::Text(text) = &mut *node.borrow_mut() else { return };
         for chunk in &mut text.chunks {
             for span in &mut chunk.spans {
-                for family in &mut span.font.families {
-                    if family.is_empty() || load(family).is_none() {
-                        if let Some(fallback) = &fallback_usvg_compatible {
-                            *family = fallback.into();
+                let Some(text) = chunk.text.get(span.start..span.end) else { continue };
+
+                let inline_families = &mut span.font.families;
+                let inline_fonts = inline_families
+                    .iter()
+                    .filter(|family| !family.is_empty())
+                    .filter_map(|family| load(family.as_str()))
+                    .collect::<EcoVec<_>>();
```

Comment:
This collect is a bit unfortunate since it goes right back to iter. I assume it's because of borrowing issues?

Reply:
Yeah, me neither. It's fine.